pkg/ringbuffer/list.go

```diff
@@ -0,0 +1,65 @@
+package ringbuffer
+
+import (
+	"sync"
+)
+
+var elemPool = &sync.Pool{New: func() interface{} { return new(element) }}
+
+type element struct {
+	next, prev *element
+	val        interface{}
+}
+
+func (e *element) reset() {
+	e.next, e.prev = nil, nil
+	e.val = nil
+}
+
+type queue struct {
+	root  element
+	count int
+}
+
+func newQueue() *queue {
+	q := new(queue)
+
+	q.root.next = &q.root
+	q.root.prev = &q.root
+	q.count = 0
+	return q
+}
+
+func (q *queue) size() int {
+	return q.count
+}
+
+func (q *queue) enqueue(val interface{}) {
+	elem := elemPool.Get().(*element)
+	elem.val = val
+
+	at := q.root.prev
+	n := at.next
```

Comment:
why we need this local var n, isn't it always q.root?

Reply:
I will make it clear.